src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_search.cs

```diff
@@ -167,38 +170,153 @@ public bool TryReadUntilAnyInternal(out ReadOnlySequence<byte> sequence, ReadOnl
             return false;
         }
 
-        public bool TryReadUntil(out ReadOnlySequence<byte> sequence, ReadOnlySpan<byte> delimiter)
+        /// <summary>
+        /// Try to read data until the given <paramref name="delimiter"/> sequence.
+        /// </summary>
+        /// <param name="sequence">The read data, if any.</param>
+        /// <param name="delimiter">The multi (T) delimiter.</param>
+        /// <param name="advancePastDelimiter">True to move past the <paramref name="delimiter"/> sequence if found.</param>
+        /// <returns>True if the <paramref name="delimiter"/> was found.</returns>
+        public unsafe bool TryReadTo(out ReadOnlySequence<T> sequence, ReadOnlySpan<T> delimiter, bool advancePastDelimiter = true)
         {
             if (delimiter.Length == 0)
             {
                 sequence = default;
                 return true;
             }
 
-            int matched = 0;
-            BufferReader copy = this;
-            SequencePosition start = Position;
-            SequencePosition end = Position;
+            BufferReader<T> copy = this;
+
+            T* b = stackalloc T[delimiter.Length];
```

Comment:
We shouldn't stackalloc without bounding the user input to some known constant.

Reply:
I'm not so sure on this one. As performance critical as this stuff is- who would want to match multiple KB? Do we support it and fall to the heap? Do we just fail past a certain point? I feel awkward penalizing everyone for an outlier, but maybe it doesn't really add up..

Reply:
I think it is a security/reliability concern. Whether we support it or not via allocating on the heap (either choice is feasible), we shouldn't be allocating an amount on the stack that is unbounded (based on user input). At least that is what was recommended to me.

cc @GrabYourPitchforks

Reply:
StackOverflow is worse than OutOfMemory I guess.

``` C#
Span<T> peekBuffer;
if (delimiter.Length * sizeof(T) < 512)
{
    T* t = stackalloc T[delimiter.Length];
    peekBuffer = new Span<T>(t, delimiter.Length);
}
else
{
    peekBuffer = new Span<T>(new T[delimiter.Length]);
}
```

Reply:
Note that I can't do the above without the unmanaged constraint.